### ISSUE TITLE
Update method on Autofac builder is obsolete

### ISSUE THREAD

**Issue by dario-l** [NancyFx/Nancy.Bootstrappers.Autofac#64]
Autofac documentation says:

> While Autofac provides an Update() method to update registrations in an existing container, for the most part it’s there for backwards-compatibility with Autofac 2.x. Where at all possible, you should avoid updating a container and instead register everything up front before building the container.

http://docs.autofac.org/en/latest/best-practices/index.html#consider-a-container-as-immutable

And they make Update obsolete:
> Containers should generally be considered immutable. Register all of your dependencies before building/resolving. If you need to change the contents of a container, you technically should rebuild the container. *This method may be removed in a future major release.*

This leads to major modifications in this bootstrapper.

**Comment by khellang** [comment]
Yes. The current design was based around the assumption that the containers could be modified after the initial registrations. This has also caused trouble with containers like SimpleInjector, which simply doesn't allow you to do this.

We've been [prototyping a bootstrapper design](https://github.com/khellang/nancy-bootstrapper-prototype) which doesn't have these assumptions and hopefully we'll be able to get that rolled into Nancy at some point in the future.

**Comment by grumpydev** [comment]
As I've said before though, that prototype doesn't cover all the use cases we cover. There's a key difference between something being immutable, and its API not allowing you to "change" it. Strings are immutable in .net, but that doesn't ,mean you can't do "mystring = mystring + otherstring", it just returns a new string instance - this is exactly how I'd expect containers to work that require internal immutability for performance.

**Comment by khellang** [comment]
@grumpydev Are you trying to say that we can't cover these use cases by having an immutable container?

**Comment by thecodejunkie** [comment]
@khellang 

> Hopefully we'll be able to get that rolled into Nancy at some point in the future.

I'll commit to work with you on this asap if you're up for it :P

**Comment by khellang** [comment]
> I'll commit to work with you on this asap if you're up for it :P

Yes! New year, new bootstrapper :grin:

**Comment by p-m-j** [comment]
Is this project dead?

Did nancy-bootstrapper-prototype get parked?

Edit: In the meantime here's a hack to allow moderately painless usage of the code from this repo with autfoac 4.9.x

**Warning** this has had all of 5 minutes testing

```
#pragma warning disable CS0618 // Type or member is obsolete (I KNOWWWWW!)
#pragma warning disable SA1402 // File may only contain a single type
public static class AutofacExtensions
{
    public static IContainer Update(this IContainer container, Action<ContainerBuilder> builderAction)
    {
        var builder = new ContainerBuilder();
        builderAction(builder);
        builder.Update(container);
        return container;
    }
}
#pragma warning restore CS0618 // Type or member is obsolete
#pragma warning restore SA1402 // File may only contain a single type

`
